subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileCompatibilityIntegrationTest.groovy

```diff
@@ -0,0 +1,587 @@
+
+        buildFile << """
+            apply plugin: 'java'
+
+            java {
+                sourceCompatibility = JavaVersion.VERSION_14
+                targetCompatibility = JavaVersion.VERSION_14
+                toolchain {
+                    languageVersion = JavaLanguageVersion.of(${javaVersion.majorVersion})
+                }
+                sourceCompatibility = null
+                targetCompatibility = null
+            }
+
+            compileJava {
+                def projectSourceCompat = project.java.sourceCompatibility
+                def projectTargetCompat = project.java.targetCompatibility
+                doLast {
+                    logger.lifecycle("project.sourceCompatibility = '\${projectSourceCompat}'")
+                    logger.lifecycle("project.targetCompatibility = '\${projectTargetCompat}'")
+                    logger.lifecycle("task.sourceCompatibility = '\$sourceCompatibility'")
+                    logger.lifecycle("task.targetCompatibility = '\$targetCompatibility'")
+                }
+            }
+        """
+
+        when:
+        withInstallations(jdk).run(":compileJava")
+
+        then:
+        outputContains("project.sourceCompatibility = '$javaVersion'")
+        outputContains("project.targetCompatibility = '$javaVersion'")
+        outputContains("task.sourceCompatibility = '$javaVersion'")
+        outputContains("task.targetCompatibility = '$javaVersion'")
+        classJavaVersion(javaClassFile("Foo.class")) == javaVersion
+    }
+
+    def "source compatibility lower than compiler version does not allow accessing newer Java language features"() {
+        def jdk = AvailableJavaHomes.getJdk(toolchain)
+
+        buildFile << """
+            apply plugin: "java"
+
+            java {
+                toolchain {
+                    languageVersion = JavaLanguageVersion.of(${jdk.javaVersion.majorVersion})
+                }
+            }
+
+            compileJava.sourceCompatibility = "$sourceCompatibility"
+        """
+
+        file("src/main/java/Parent.java") << """
+            // Sealed classes and interfaces are only available in Java 17
+            public sealed interface Parent permits Parent.Child {
+                public static record Child(String name) implements Parent {}
+            }
+        """
+
+        when:
+        if (fails) {
+            withInstallations(jdk).fails(":compileJava")
+        } else {
+            withInstallations(jdk).succeeds(":compileJava")
+        }
+
+        then:
+        if (fails) {
+            failure.assertHasErrorOutput("Parent.java:3: error: sealed classes are not supported in -source 11")
+            javaClassFile("Parent.class").assertDoesNotExist()
+        } else {
+            classJavaVersion(javaClassFile("Parent.class")) == JavaVersion.VERSION_17
+        }
+
+        where:
+        sourceCompatibility    | toolchain              | fails
+        JavaVersion.VERSION_17 | JavaVersion.VERSION_17 | false
```

Comment:
nitpick: I probably wouldn't test the positive case here, especially since it makes the test kind of convoluted.  If  this coverage is important, I'd use a separate test for it.